### ISSUE TITLE
export 'sanitize' (imported as 'sanitize') was not found in 'dompurify' (possible exports: default)

### ISSUE THREAD

**Issue by ngtrthinh169TPC** [cure53/DOMPurify#1034]
> This issue proposes a bug which `dompurify`'s `typescript` is messed up.

### Background & Context

DefinitelyTyped has removed `@types/dompurify` 2 days ago (https://github.com/DefinitelyTyped/DefinitelyTyped/pull/71153), which renders the old version of `dompurify` not buildable within projects that uses `typescript`. We were forced to update the version to ^3.2.0. After updating, several breaking changes happened.

#### Input

My old import:

```
import { sanitize } from "dompurify";
```

#### Given output

When build, a warning shows up in the console:
`export 'sanitize' (imported as 'sanitize') was not found in 'dompurify' (possible exports: default)`

#### Expected output

The old code should not break.

### Feature

Keep the export as it used to be.

Or just let us know how to migrate this.

**Comment by ngtrthinh169TPC** [comment]
Hey to add some more contexts I tried

```
import DOMPurify from "dompurify";
console.log(DOMPurify, typeof DOMPurify);
```

and it prints `root => createDOMPurify(root) 'function'`, which is not an object and not the same as the defined types.

Also I'm using v3.2.1

**Comment by ngtrthinh169TPC** [comment]
Hey I've found a work-around for this by using default export 

```
import DOMPurify from "dompurify";
```

and use `DOMPurify.sanitize`.

Not quite the prettiest piece of code since it triggers some eslint errors of "no-undef", but after some researches I decide to turn off that eslint rule and settle with the default export (ref: https://github.com/typescript-eslint/typescript-eslint/issues/662). For that matter I'll close this issue here.

Still, I think removing `@types/dompurify` immediately is a bad move for general users.

**Comment by MrChocolatine** [comment]
@ngtrthinh169TPC please reopen this ticket, this was and is still an issue and should be marked as "Completed".

The symbol `sanitize` is still visible but importing results in this error.
Somehow went wrong with the recent updates.

cc @cure53

**Comment by ngtrthinh169TPC** [comment]
I wasn't sure if that was my fault or was it the new version, since the export code doesn't seem to be changed. But seeing there's multiple people out there with the same problem, I think it's safe to say we should at least revert the removal of `@types/dompurify` until things are stable enough.

**Comment by cure53** [comment]
Heya, we currently have no plans to revert anything, or dig deeper on this issue, as it seems to work well for a majority of folks. 

Please check what's causing the issue on your end. If anything is the outcome that we can do to fix it, happy to receive a PR.

**Comment by MrChocolatine** [comment]
> Heya, we currently have no plans to revert anything, or dig deeper on this issue, as it seems to work well for a majority of folks.

And... I was expecting this. I read a lot of issues on this repo and 98% of the time it's the kind of answer always given.
"Not our problem, we won't do anything, close the issue without any (or almost) further discussion"

What a way to maintain a library and exchange with your users.

**Comment by cure53** [comment]
I am sorry you feel frustrated, but this is really an issue you need to figure out yourself. 

If we change something this direction, people will have issues, if we change something the other direction, other people will have issues. We use definitely types? People cry. We use out own types? People cry.

If you do however find something that we actually did wrong and have a solution, we are super-open to looking into that, accepting a PR and making it work. Everything else really is your problem and way outside what we can do in the little bit of spare time we have to maintain this project.

**Comment by Windmolders** [comment]
There is always a small way around by doing the following. Which is extra code yes, but easier then editing the whole code:
I came here to see if there was another solution or something.

`import DOMPurify from 'dompurify'

const { sanitize } = DOMPurify`

**Comment by steveschmitt** [comment]
I ended up doing this.  No compile problems, but not pretty.  Maybe there is a better way?

```ts
import * as dompurify from 'dompurify';
const purifier = dompurify['default'] as dompurify.DOMPurify;

export const purify = (s: string, cfg?: dompurify.Config) => {
  return purifier.sanitize(s, cfg) as string;
}
```

**Comment by carlosmintfan** [comment]
`import createDompurify, { Config, DOMPurifyI, SanitizeAttributeHookEvent, SanitizeElementHookEvent } from 'dompurify';` broke for me. `DOMPurifyI, SanitizeAttributeHookEvent, SanitizeElementHookEvent` are not exported by `dompurify`.

**Comment by alhawks** [comment]
I would also like to see this issue reopened. The suggested workaround of importing DOMPurify instead of ` { sanitize } from dompurify` is easy enough to implement, but that's not the biggest problem. The biggest problem is that it now makes version 3.2 of this package incompatible with 3.1 and 3.0. I have a dependency that requires `^3.0.1`, but that doesn't work anymore if I provide a 3.2 version, since it breaks the import in my dependency.

In my opinion, this should be addressed to make sure it's backwards compatible, or make this a version 4, since this seems to be a breaking change.

**Comment by cure53** [comment]
@alhawks If the dependency needs 3.0.1 to work, which is a bypassable version, then that dependency likely introduces XSS and should be either fixed or dropped, no?

**Comment by rohDevellopper** [comment]
It worked for me like this:

`import dompurify from "isomorphic-dompurify";
export function sanitizeInput(input: string): string {
  return dompurify.sanitize(input.trim())
}`

**Comment by ngtrthinh169TPC** [comment]
@cure53 

> If the dependency needs 3.0.1 to work, which is a bypassable version, then that dependency likely introduces XSS and should be either fixed or dropped, no?

I think not. A "bypassable" version still works, just not perfectly. Yes, it might need to be fixed, but not all things have its maintainer. No, most of the time, dropping is not a viable choice for most of us. We'd rather use a vulnerable software than not having anything to use at all since security is usually not the most important thing.

May I ask your reasoning behind not supporting backward compatibility? What's the downside of it? Kinda curious about that perspective since that does look like an inconventional choice, at least on this thread.

**Comment by carlosmintfan** [comment]
If Dompurify was fully semver-compliant, it would be okay for the dependency to require ^3.0.1 as the ^ also allows any newer version as long as the major version stays the same.
@alhawks As a workaround, you could perhaps patch the import in your dependency (using something like the patch-package npm package).
I also agree that not being semver compliant is not cool...though I didn't dig deeply into this issue so I don't know what makes cure53 not want to fix it.

**Comment by cure53** [comment]
> May I ask your reasoning behind not supporting backward compatibility? What's the downside of it? Kinda curious about that perspective since that does look like an inconventional choice, at least on this thread.

It's not an unconventional choice but the only one. The sanitizer has one job, prevent XSS. If a version has a bypass, it should not be used anymore, because it doesn't do the job it is supposed to do. 

Backwards compatibly is a goal that's quite orthogonal to that, and therefore isn't and won't be a priority.

**Comment by cure53** [comment]
> If Dompurify was fully semver-compliant

While we try our best to use version numbers that make sense, we also don't follow the goal of being fully semver compliant, because it's close to impossible. Any change might be a breaking change if folks expect the same sanitizer results and have unit tests for that.

**Comment by carlosmintfan** [comment]
> I ended up doing this.  No compile problems, but not pretty.  Maybe there is a better way?
> 
> ```ts
> import * as dompurify from 'dompurify';
> const purifier = dompurify['default'] as dompurify.DOMPurify;
> 
> export const purify = (s: string, cfg?: dompurify.Config) => {
>   return purifier.sanitize(s, cfg) as string;
> }
> ```

@steveschmitt Your two first lines can be simplified to `import purifier from 'dompurify'` as this will already bring the default export into a variable named purifier. Then you can also get rid of the purify function and just use purifier.sanitize everywhere I think.

**Comment by carlosmintfan** [comment]
I've perhaps found the cause of the issue; would be a quite obscure one and understandable if @cure53 doesn't want to waste time on it. I'm not sure if I'm right with my theory however so I'll do some testing the next days and tell you my theory if I can confirm it.

**Comment by cure53** [comment]
@carlosmintfan We're always happy to review and accept PRs that make things better for folks out there, but we are also aware that one person's fix is another person's breakage in this realm 😅 

So, if you have a finding that you wanna share and something that can be improved, I'd be all ears.

**Comment by alhawks** [comment]
Thanks @carlosmintfan for digging into that. What you're saying makes a lot of sense.

I am also sympathetic to what @cure53 about the whole point of this package being for security, so it's not good to depend on a package that is using an older and vulnerable version of this package. I ended up working with the owner of that dependency to update their version of dompurify.

I will say that I have a different view on semver compliance than @cure53. The thing that matters to me is that the "contract" is semver compliant. As long as I can keep calling the package in the same way, I wouldn't consider that a breaking change even if the output of a particular usage did change. I've always viewed it as "can I keep calling this package in the same way to similar effect". If that's true, then I wouldn't consider it a breaking change even if the output was different.

**Comment by cure53** [comment]
> I ended up working with the owner of that dependency to update their version of dompurify.

This is amazing, awesome work 😄 

> If that's true, then I wouldn't consider it a breaking change even if the output was different.

I fully agree. And frankly, we haven't changed our API at all for a very long time. And when we did, we used the right version number change. At least that is what I want to believe 😅 

The breaking changes, as far as I can see, came from changes to other parts than the actual library... but I do agree, it's still annoying for everyone (including us) - I am just not sure if stricter or larger-scope semver would have helped here.
